File: Data Structures & Algorithms/sliding-window-maximum/submission-8.py
import heapq
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # left, right 
        # store a heap with indices as second value
        if k == 1:
            return nums
        left = 0
        target = []
        ans = []
        for i in range(len(nums)):
            heapq.heappush_max(target, (nums[i], i))
            if i - left + 1 == k:
                #print(target)
                #print(cur)
                while target[0][1] < left or target[0][1] > i:
                    heapq.heappop_max(target)
                ans.append(target[0][0])
                left += 1

        return ans
            
            
            

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-8.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];

        // Left products
        int left = 1;
        for (int i = 0; i < nums.length; i++) {
            result[i] = left;
            left *= nums[i];
        }

        // Right products
        int right = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            result[i] *= right;
            right *= nums[i];
        }

        return result;
    }
}
